Tasks/Npm/npmtask.ts

```diff
@@ -1,27 +1,189 @@
+        if (tl.getVariable('system.debug') != undefined && tl.getVariable('system.debug').toLowerCase() === 'false') {
+            debugLog = false;
+        }
 
-var cwd = tl.getPathInput('cwd', true, false);
-tl.mkdirP(cwd);
-tl.cd(cwd);
+        try{
+            await runNpmAuthHelperAsync(getNpmAuthHelperRunner(npmrcPath, tempNpmrcPath, debugLog));
 
-var command = tl.getInput('command', true);
-if (command.indexOf(' ') >= 0) {
-	tl.setResult(tl.TaskResult.Failed, tl.loc("InvalidCommand"));
+            // set required environment variables for npm execution
+            var npmExecOptions: trm.IExecOptions = {};
+            npmExecOptions.env = process.env;
+            npmExecOptions.env['npm_config_userconfig'] = tempNpmrcPath;
+            if(debugLog) {
+                npmExecOptions.env['npm_config_loglevel'] =  'verbose';
+            }
+
+            await tryRunNpmConfigAsync(getNpmConfigRunner(debugLog), npmExecOptions);
+            var code : number =  await runNpmCommandAsync(npmRunner, npmExecOptions);
+            tl.setResult(code, tl.loc('NpmReturnCode', code));
+        } catch (err) {
+            cleanUpTempNpmrcPath(tempNpmrcPath);
+            tl.setResult(tl.TaskResult.Failed, tl.loc('NpmFailed', err.message));
+        }
+    }
+}
+
+async function runNpmAuthHelperAsync(npmAuthRunner: trm.ToolRunner) : Promise<number> {
+    try{
+        var execOptions : trm.IExecOptions = {};
+        execOptions.env = process.env || getBuildCredProviderEnv();
+        var code : number = await npmAuthRunner.exec(execOptions);
+        tl.debug('Authentication succeeded with code: ' + code);
+        return Q(code);
+    } catch (err) {
+        tl.debug(tl.loc('NpmAuthFailed', err.message));
+        throw err;
+    }
+}
+
+async function tryRunNpmConfigAsync(npmConfigRunner: trm.ToolRunner, execOptions : trm.IExecOptions) {
+    try {
+        var code = await npmConfigRunner.exec(execOptions);
+    } catch (err) {
+        // do not throw on this failure.
+        tl.warning(tl.loc('NpmConfigFailed', err.Message));
+    }
+}
+
+async function runNpmCommandAsync(npmCommandRunner: trm.ToolRunner, execOptions : trm.IExecOptions) : Promise<number> {
+    try{
+        var code: number = await npmCommandRunner.exec(execOptions);
+        tl.debug('Npm command succeeded with code: ' + code);
+        return Q(code);
+    } catch (err) {
+        tl.debug(tl.loc('NpmFailed', err.message));
+        throw err;
+    }
 }
-npm.arg(command);
 
-npm.argString(tl.getInput('arguments', false));
+function shouldUseDeprecatedTask() : boolean {
+    return tl.getVariable('USE_DEPRECATED_TASK_VERSION') != undefined && tl.getVariable('USE_DEPRECATED_TASK_VERSION').toLowerCase() === 'true';
+}
+
+function getNpmAuthHelperRunner(npmrcPath: string, tempUserNpmrcPath: string, includeDebugLogs: boolean): trm.ToolRunner {
+    var npmAuthHelper = tl.createToolRunner(path.join(__dirname, 'Npm/vsts-npm-auth/bin/vsts-npm-auth.exe'));
```

Comment:
I felt it would be better to use the explicit one. NuGet does the same. Let's talk about the L0's if there's a better way to mock with the current behavior.